src/circuit/gadgets/constraints.rs

```diff
@@ -427,6 +464,148 @@ pub fn must_be_simple_bit(x: &Boolean) -> AllocatedBit {
     }
 }
 
+// Enforce div and mod operation for U64. We need to show that
+// arg1 = q * arg2 + r, such that 0 <= r < arg2.
+pub fn enforce_u64_div_mod<F: LurkField, CS: ConstraintSystem<F>>(
+    mut cs: CS,
+    g: &GlobalAllocations<F>,
+    cond: Boolean,
+    arg1: &AllocatedPtr<F>,
+    arg2: &AllocatedPtr<F>,
+    store: &Store<F>,
+) -> Result<(AllocatedPtr<F>, AllocatedPtr<F>), SynthesisError> {
+    let arg1_u64 = match arg1.hash().get_value() {
+        Some(v) => v.to_u64_unchecked(),
+        None => 0, // Blank and Dummy
+    };
+    let arg2_u64 = match arg2.hash().get_value() {
+        Some(v) => v.to_u64_unchecked(),
+        None => 0, // Blank and Dummy
+    };
+
+    let (q, r) = if arg2_u64 != 0 {
+        (arg1_u64 / arg2_u64, arg1_u64 % arg2_u64)
+    } else {
+        (0, 0)
+    };
+
+    let r_ptr = store.get_u64(r);
+    let alloc_r =
+        AllocatedPtr::alloc_ptr(&mut cs.namespace(|| "u64 remainder"), store, || Ok(&r_ptr))?;
+
+    let q_ptr = store.get_u64(q);
+    let alloc_q =
+        AllocatedPtr::alloc_ptr(&mut cs.namespace(|| "u64 quotient"), store, || Ok(&q_ptr))?;
+
+    let arg1_u64_ptr = store.get_u64(arg1_u64);
+    let alloc_arg1 = AllocatedPtr::alloc_ptr(&mut cs.namespace(|| "alloc arg1"), store, || {
+        Ok(&arg1_u64_ptr)
+    })?;
```

Comment:
What is the purpose of the whole song-and-dance to created an `AllocatedPtr` here? Why can't we just create an `AllocatedNum` direcdtly from `arg1_u64`. As far as I can tell, the tag part is never used — except that you enforce the constraint that it be 'correct' — but correct for what purpose?

Reply:
fixed